crypto/evp/kdf_lib.c

```diff
@@ -109,13 +109,18 @@ void EVP_KDF_reset(EVP_KDF_CTX *ctx)
 
 size_t EVP_KDF_size(EVP_KDF_CTX *ctx)
 {
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    size_t s;
+
     if (ctx == NULL)
         return 0;
 
-    if (ctx->meth->size == NULL)
-        return SIZE_MAX;
-
-    return ctx->meth->size(ctx->data);
+    if (ctx->meth->get_ctx_params != NULL) {
+        *params = OSSL_PARAM_construct_size_t(OSSL_KDF_PARAM_SIZE, &s);
+        if (ctx->meth->get_ctx_params(ctx, params) > 0)
+            return s;
```

Comment:
Nit: `get_ctx_params` is supposesd to return a boolean, i.e. 0 or 1.  So `> 0` is unnecessary...

Also, is the size variable per operation, or is it implementation specific?  This might vary of course...  I chose to use both `get_ctx_params` and `get_params` in the MAC move.  Totally your call how you want to do things, just wanted to give you food for thought.

Reply:
Most of the ones with variable size, depend on the underlying digest/mac for their size.
It doesn't seem appropriate to use `get_params`.

I guess, I could try `get_ctx_params` first and fall back to `get_params` if it isn't implemented or fails.

Reply:
Fixed the `> 0` nit.

Reply:
> I guess, I could try `get_ctx_params` first and fall back to `get_params` if it isn't implemented or fails.

Yes.  That's exactly what is done by `EVP_MAC_size()`

Reply:
I still wonder -- currently only one I think returns a constant.  I'll check how many don't have such a function.

Reply:
The more I think about this, the less sure I am that it makes sense.  Wouldn't it mean two public API's `EVP_KDF_size` and `EVP_KDF_CTX_size`?  One of which will mostly return zero.  It feels complicating without much benefit.

Reply:
I don't see that it needs to translate in that kind of dual API is particularly interesting...  I've chosen to have a `EVP_MAC_size()` that takes an `EVP_MAC_CTX` pointer, but that calls the `get_ctx_params` dipatched function first, and `get_params` as a fallback.

Reply:
Changed to check both calls.  I've not implemented a `get_params` for any algorithm.